### PR TITLE
add output of dissolved solvent rswsol

### PR DESCRIPTION
**Pull request by totto82**
Depends on https://github.com/OPM/opm-common/pull/3764 and https://github.com/OPM/opm-models/pull/851

### CODE REVIEW

File: ebos/eclgenericoutputblackoilmodule.cc [hunk 1]
```diff
@@ -633,6 +634,12 @@ setRestart(const data::Solution& sol,
         so -= sSol_[elemIdx];
     }
 
+    if (!rswSol_.empty()) {
+        if (sol.has("RSWSOL"))
+            rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
```
Comment (flammmes) [inline ebos/eclgenericoutputblackoilmodule.cc:639]: I tried to build this and got an error.
[  1%] Building CXX object CMakeFiles/opmsimulators.dir/ebos/eclgenericoutputblackoilmodule.cc.o
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc: In member function ‘void Opm::EclGenericOutputBlackoilModule<FluidSystem, Scalar>::setRestart(const Opm::data::Solution&, unsigned int, unsigned int)’:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: error: no matching function for call to ‘Opm::data::Solution::data(const char [7]) const’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
In file included from /root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:31:
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:51:25: note: candidate: ‘template<class T> std::vector<Scalar>& Opm::data::Solution::data(const string&)’
   51 |         std::vector<T>& data(const std::string& );
      |                         ^~~~
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:51:25: note:   template argument deduction/substitution failed:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: note:   couldn’t deduce template parameter ‘T’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
In file included from /root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:31:
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:54:31: note: candidate: ‘template<class T> const std::vector<Scalar>& Opm::data::Solution::data(const string&) const’
   54 |         const std::vector<T>& data(const std::string& ) const;
      |                               ^~~~
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:54:31: note:   template argument deduction/substitution failed:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: note:   couldn’t deduce template parameter ‘T’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
make[2]: *** [CMakeFiles/opmsimulators.dir/build.make:104: CMakeFiles/opmsimulators.dir/ebos/eclgenericoutputblackoilmodule.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:338: CMakeFiles/opmsimulators.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

So I changed it to 
rswSol_[elemIdx] = sol.data<Scalar>("RSWSOL")[globalDofIndex];
and now it builds fine.

Comment (flammmes) [inline ebos/eclgenericoutputblackoilmodule.cc:639]: I tried to build this but got an error.
[  1%] Building CXX object CMakeFiles/opmsimulators.dir/ebos/eclgenericoutputblackoilmodule.cc.o
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc: In member function ‘void Opm::EclGenericOutputBlackoilModule<FluidSystem, Scalar>::setRestart(const Opm::data::Solution&, unsigned int, unsigned int)’:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: error: no matching function for call to ‘Opm::data::Solution::data(const char [7]) const’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
In file included from /root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:31:
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:51:25: note: candidate: ‘template<class T> std::vector<Scalar>& Opm::data::Solution::data(const string&)’
   51 |         std::vector<T>& data(const std::string& );
      |                         ^~~~
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:51:25: note:   template argument deduction/substitution failed:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: note:   couldn’t deduce template parameter ‘T’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
In file included from /root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:31:
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:54:31: note: candidate: ‘template<class T> const std::vector<Scalar>& Opm::data::Solution::data(const string&) const’
   54 |         const std::vector<T>& data(const std::string& ) const;
      |                               ^~~~
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:54:31: note:   template argument deduction/substitution failed:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: note:   couldn’t deduce template parameter ‘T’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
make[2]: *** [CMakeFiles/opmsimulators.dir/build.make:104: CMakeFiles/opmsimulators.dir/ebos/eclgenericoutputblackoilmodule.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:338: CMakeFiles/opmsimulators.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
root@MSI:~/opm_beta/opm-simulators/build# make
[  1%] Updating version information
[  1%] Built target update-version
Consolidate compiler generated dependencies of target opmsimulators
[  1%] Building CXX object CMakeFiles/opmsimulators.dir/ebos/eclgenericoutputblackoilmodule.cc.o
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc: In member function ‘void Opm::EclGenericOutputBlackoilModule<FluidSystem, Scalar>::setRestart(const Opm::data::Solution&, unsigned int, unsigned int)’:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: error: no matching function for call to ‘Opm::data::Solution::data(const char [7]) const’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
In file included from /root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:31:
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:51:25: note: candidate: ‘template<class T> std::vector<Scalar>& Opm::data::Solution::data(const string&)’
   51 |         std::vector<T>& data(const std::string& );
      |                         ^~~~
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:51:25: note:   template argument deduction/substitution failed:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: note:   couldn’t deduce template parameter ‘T’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
In file included from /root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:31:
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:54:31: note: candidate: ‘template<class T> const std::vector<Scalar>& Opm::data::Solution::data(const string&) const’
   54 |         const std::vector<T>& data(const std::string& ) const;
      |                               ^~~~
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:54:31: note:   template argument deduction/substitution failed:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: note:   couldn’t deduce template parameter ‘T’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
make[2]: *** [CMakeFiles/opmsimulators.dir/build.make:104: CMakeFiles/opmsimulators.dir/ebos/eclgenericoutputblackoilmodule.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:338: CMakeFiles/opmsimulators.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

So I changed it to `rswSol_[elemIdx] = sol.data<Scalar>("RSWSOL")[globalDofIndex];
and now it builds

Comment (flammmes) [inline ebos/eclgenericoutputblackoilmodule.cc:639]: I tried to build this but got an error.
[  1%] Building CXX object CMakeFiles/opmsimulators.dir/ebos/eclgenericoutputblackoilmodule.cc.o
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc: In member function ‘void Opm::EclGenericOutputBlackoilModule<FluidSystem, Scalar>::setRestart(const Opm::data::Solution&, unsigned int, unsigned int)’:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: error: no matching function for call to ‘Opm::data::Solution::data(const char [7]) const’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
In file included from /root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:31:
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:51:25: note: candidate: ‘template<class T> std::vector<Scalar>& Opm::data::Solution::data(const string&)’
   51 |         std::vector<T>& data(const std::string& );
      |                         ^~~~
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:51:25: note:   template argument deduction/substitution failed:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: note:   couldn’t deduce template parameter ‘T’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
In file included from /root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:31:
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:54:31: note: candidate: ‘template<class T> const std::vector<Scalar>& Opm::data::Solution::data(const string&) const’
   54 |         const std::vector<T>& data(const std::string& ) const;
      |                               ^~~~
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:54:31: note:   template argument deduction/substitution failed:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: note:   couldn’t deduce template parameter ‘T’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
make[2]: *** [CMakeFiles/opmsimulators.dir/build.make:104: CMakeFiles/opmsimulators.dir/ebos/eclgenericoutputblackoilmodule.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:338: CMakeFiles/opmsimulators.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
root@MSI:~/opm_beta/opm-simulators/build# make
[  1%] Updating version information
[  1%] Built target update-version
Consolidate compiler generated dependencies of target opmsimulators
[  1%] Building CXX object CMakeFiles/opmsimulators.dir/ebos/eclgenericoutputblackoilmodule.cc.o
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc: In member function ‘void Opm::EclGenericOutputBlackoilModule<FluidSystem, Scalar>::setRestart(const Opm::data::Solution&, unsigned int, unsigned int)’:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: error: no matching function for call to ‘Opm::data::Solution::data(const char [7]) const’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
In file included from /root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:31:
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:51:25: note: candidate: ‘template<class T> std::vector<Scalar>& Opm::data::Solution::data(const string&)’
   51 |         std::vector<T>& data(const std::string& );
      |                         ^~~~
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:51:25: note:   template argument deduction/substitution failed:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: note:   couldn’t deduce template parameter ‘T’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
In file included from /root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:31:
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:54:31: note: candidate: ‘template<class T> const std::vector<Scalar>& Opm::data::Solution::data(const string&) const’
   54 |         const std::vector<T>& data(const std::string& ) const;
      |                               ^~~~
/root/opm_beta/opm-common/opm/output/data/Solution.hpp:54:31: note:   template argument deduction/substitution failed:
/root/opm_beta/opm-simulators/ebos/eclgenericoutputblackoilmodule.cc:658:40: note:   couldn’t deduce template parameter ‘T’
  658 |             rswSol_[elemIdx] = sol.data("RSWSOL")[globalDofIndex];
      |                                ~~~~~~~~^~~~~~~~~~
make[2]: *** [CMakeFiles/opmsimulators.dir/build.make:104: CMakeFiles/opmsimulators.dir/ebos/eclgenericoutputblackoilmodule.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:338: CMakeFiles/opmsimulators.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

So I changed it to
image

and now it builds. Sorry for uploading an image but github removed the angle brackets everytime I commented.

Comment (totto82) [reply ebos/eclgenericoutputblackoilmodule.cc:639]: Thanks. I will update the PR accordingly. I will prioritize to get the PR into master to make it easier to use.

Comment (akva2) [reply ebos/eclgenericoutputblackoilmodule.cc]: for future reference: use html tags for angular brackets, ie sol.data & lt ;Scalar & gt; (without the spaces) or wrap it in a code block,
```
sol.data<Scalar>
```

File: ebos/eclgenericoutputblackoilmodule.cc [hunk 2]
```diff
@@ -966,6 +973,7 @@ doAllocBuffers(const unsigned bufferSize,
 
     if (enableSolvent_) {
         sSol_.resize(bufferSize, 0.0);
+        rswSol_.resize(bufferSize, 0.0);
```
Comment (svenn-t) [inline ebos/eclgenericoutputblackoilmodule.cc:976]: RSWSOL will be outputted even when DISGASW is disabled

**Comment by svenn-t** [comment]
I have tested H2SOL with different combinations and they all seems to work well, except for the combination H2SOL, DISGASW and MISC table (with other necessary modifications). The simulation finishes but gave NaN values for SOIL. Test case attached if you want to try: 
[h2sol.zip](https://github.com/OPM/opm-simulators/files/13686138/h2sol.zip)

**Comment by totto82** [comment]
@svenn-t has communicated offline that with the latest fix all models runs fine. Will start the merging process.

**Pull request merged**